pkg/util/mount/mount_linux.go

```diff
@@ -687,102 +684,137 @@ func doMakeRShared(path string, mountInfoFilename string) error {
 		}
 		// Ignore ErrorNotExist: the file/directory will be created below if it does not exist yet.
 		notMount = true
 	}
 	if !notMount {
 		// It's already mounted
 		glog.V(5).Infof("Skipping bind-mounting subpath %s: already mounted", bindPathTarget)
-		success = true
-		return bindPathTarget, nil
+		return true, bindPathTarget, nil
+	}
+
+	// bindPathTarget is in /var/lib/kubelet and thus reachable without any
+	// translation even to containerized kubelet.
+	bindParent := filepath.Dir(bindPathTarget)
+	err = os.MkdirAll(bindParent, 0750)
+	if err != nil && !os.IsExist(err) {
+		return false, "", fmt.Errorf("error creating directory %s: %s", bindParent, err)
 	}
 
-	// Create target of the bind mount. A directory for directories, empty file
-	// for everything else.
 	t, err := os.Lstat(subpath.Path)
 	if err != nil {
-		return "", fmt.Errorf("lstat %s failed: %s", subpath.Path, err)
+		return false, "", fmt.Errorf("lstat %s failed: %s", subpath.Path, err)
 	}
+
 	if t.Mode()&os.ModeDir > 0 {
 		if err = os.Mkdir(bindPathTarget, 0750); err != nil && !os.IsExist(err) {
-			return "", fmt.Errorf("error creating directory %s: %s", bindPathTarget, err)
+			return false, "", fmt.Errorf("error creating directory %s: %s", bindPathTarget, err)
 		}
 	} else {
-		// "/bin/touch <bindDir>".
+		// "/bin/touch <bindPathTarget>".
 		// A file is enough for all possible targets (symlink, device, pipe,
 		// socket, ...), bind-mounting them into a file correctly changes type
 		// of the target file.
 		if err = ioutil.WriteFile(bindPathTarget, []byte{}, 0640); err != nil {
-			return "", fmt.Errorf("error creating file %s: %s", bindPathTarget, err)
+			return false, "", fmt.Errorf("error creating file %s: %s", bindPathTarget, err)
 		}
 	}
+	return false, bindPathTarget, nil
+}
+
+func getSubpathBindTarget(subpath Subpath) string {
+	// containerName is DNS label, i.e. safe as a directory name.
+	return filepath.Join(subpath.PodDir, containerSubPathDirectoryName, subpath.VolumeName, subpath.ContainerName, strconv.Itoa(subpath.VolumeMountIndex))
+}
+
+func doBindSubPath(mounter Interface, subpath Subpath) (hostPath string, err error) {
+	// Linux, kubelet runs on the host:
+	// - safely open the subpath
+	// - bind-mount /proc/<pid of kubelet>/fd/<fd> to subpath target
+	// User can't change /proc/<pid of kubelet>/fd/<fd> to point to a bad place.
+
+	// Evaluate all symlinks here once for all subsequent functions.
+	newVolumePath, err := filepath.EvalSymlinks(subpath.VolumePath)
+	if err != nil {
+		return "", fmt.Errorf("error resolving symlinks in %q: %v", subpath.VolumePath, err)
+	}
+	newPath, err := filepath.EvalSymlinks(subpath.Path)
+	if err != nil {
+		return "", fmt.Errorf("error resolving symlinks in %q: %v", subpath.Path, err)
+	}
+	glog.V(5).Infof("doBindSubPath %q (%q) for volumepath %q", subpath.Path, newPath, subpath.VolumePath)
+	subpath.VolumePath = filepath.Clean(newVolumePath)
```

Comment:
Clean is not needed here since it's done in doSafeOpen()?

Reply:
fixed